File: tests/test_module_sdk_abi.py
"""I qualify metadata ownership and actual cache fingerprints without providers."""
import copy
import json
import os
from pathlib import Path
import shlex
import sys
import tempfile
import unittest
from tests.test_file_cyclic import FileCyclic
ROOT = Path(__file__).resolve().parents[1]

class ModuleSdkAbi(unittest.TestCase):
    command = FileCyclic.command

    def setUp(self):
        self.artifacts = Path(tempfile.mkdtemp(prefix='nano-sdk-abi-metadata-'))
        self.serial = 0
        self.probe = Path(os.environ['SDK_ABI_PROBE'])
        self.module = self.artifacts / 'provider'
        self.module.mkdir()

    def run_checked(self, args):
        self.serial += 1
        return self.command(str(self.serial), list(map(str, args)))

    def schema(self):
        return {'version': 1, 'target': 'fixture-host',
                'types': [{'semantic': 'int', 'c_type': 'int64_t', 'kind': 'int'},
                          {'semantic': 'Packet', 'c_type': 'Packet', 'kind': 'record', 'fields': ['x', 'owner']},
                          {'semantic': 'array<Packet>', 'c_type': 'DynArray', 'kind': 'array', 'abi': 'dyn_array_v2', 'element': 1}],
                'functions': [{'name': 'read', 'symbol': 'provider_read', 'parameters': [2], 'result': 1}]}

    def probe_metadata(self, schema=None, raw=None, refused=False):
        metadata = {'name': 'fixture', 'version': '1', 'c_sources': []}
        if schema is not None:
            metadata['typed_abi'] = schema
        (self.module / 'module.json').write_text(raw if raw is not None else json.dumps(metadata))
        if refused:
            # The outer owning command supervises this child in the same group.
            # Sanitizer diagnostics cannot be mistaken for an expected refusal.
            script = ('import subprocess,sys; r=subprocess.run(sys.argv[1:],capture_output=True); '
                      'sys.stdout.buffer.write(r.stdout); sys.stderr.buffer.write(r.stderr); '
                      'assert r.returncode==1; '
                      'assert not any(x in r.stderr for x in (b"Sanitizer",b"runtime error:"))')
            self.run_checked([sys.executable, '-c', script, self.probe, 'typed-abi', self.module])
            return None
        output = self.run_checked([self.probe, 'typed-abi', self.module]).decode().splitlines()
        self.assertEqual(len(output), 2)
        exact, absent = map(int, output[1].split())
        self.assertGreater(exact, 0)
        self.assertGreater(absent, 0)
        return json.loads(output[0]), exact, absent

    def test_owned_parser_and_allocation_rollback(self):
        compiler = shlex.split(os.environ['SDK_ABI_CC'])
        flags = shlex.split(os.environ['SDK_ABI_CFLAGS'])
        flags = ['-std=c11' if flag.startswith('-std=') else flag for flag in flags]
        output = self.artifacts / 'owned'
        self.run_checked([*compiler, *flags, '-Isrc', 'tests/test_module_sdk_abi.c',
                          os.environ['SDK_ABI_CJSON'], *shlex.split(os.environ['SDK_ABI_LDFLAGS']), '-o', output])
        self.assertIn(b'owned typed ABI parsing, roundtrip and allocation rollback', self.run_checked([output]))

    def test_actual_fingerprint_and_refusals(self):
        absent, old, baseline = self.probe_metadata()
        self.assertIsNone(absent)
        self.assertEqual(old, baseline)
        schema = self.schema()
        canonical, first, old = self.probe_metadata(schema)
        self.assertEqual(canonical, schema)
        self.assertNotEqual(first, old)
        self.assertEqual(old, baseline)
        reordered = {key: schema[key] for key in reversed(schema)}
        reordered['types'] = [{key: row[key] for key in reversed(row)} for row in schema['types']]
        self.assertEqual(self.probe_metadata(reordered), (canonical, first, baseline))
        self.assertEqual(self.probe_metadata(canonical), (canonical, first, baseline))
        changes = [lambda s: s.update(target='another-host'),
                   lambda s: s['types'][1].update(c_type='OtherPacket'),
                   lambda s: s['types'][1]['fields'].reverse(),
                   lambda s: s['functions'][0].update(symbol='other_read'),
                   lambda s: s['functions'][0].update(result=0)]
        for change in changes:
            changed = copy.deepcopy(schema)
            change(changed)
            _, new, same_old = self.probe_metadata(changed)
            self.assertNotEqual(new, first)
            self.assertEqual(same_old, baseline)
        bad = [lambda s: s.update(version=2),
               lambda s: s.update(unknown=True),
               lambda s: s['types'][1].update(c_type='Packet *'),
               lambda s: s['types'][1].update(fields=['x', 'x']),
               lambda s: s['types'][1].update(fields=['x; injected']),
               lambda s: s['types'][2].update(element=3),
               lambda s: s['types'][2].update(element=1.5),
               lambda s: s['types'][2].update(abi='unknown_abi'),
               lambda s: s['functions'][0].update(result=-1),
               lambda s: s['functions'][0].update(parameters=[3]),
               lambda s: s['functions'].append(copy.deepcopy(s['functions'][0])),
               lambda s: s['types'].append(copy.deepcopy(s['types'][0])),
               lambda s: s.update(target='x' * 4097),
               lambda s: s.update(types=[{'semantic': str(i), 'c_type': 'T', 'kind': 'int'} for i in range(4097)])]
        for change in bad:
            changed = copy.deepcopy(schema)
            change(changed)
            self.probe_metadata(changed, refused=True)
        encoded = json.dumps(schema)
        self.probe_metadata(raw='{"typed_abi":' + encoded + ',"typed_abi":' + encoded + '}', refused=True)
        self.probe_metadata(raw='{"typed_abi":' + encoded.replace('"version": 1', '"version": 1,"version": 1') + '}', refused=True)
        self.probe_metadata(raw='{"typed_abi":' + encoded.replace('fixture-host', 'fixture\\u0000host') + '}', refused=True)

if __name__ == '__main__':
    unittest.main()

File: tests/test_module_sdk_abi.c
/* I qualify owned parsing only; no generated adapter or provider is invoked. */
#ifdef NDEBUG
#error I require active ownership assertions
#endif
#include "module_builder.h"
#include "cJSON.h"
#include <assert.h>
#include <stdint.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
static size_t calls,live,fail_at=SIZE_MAX;
static bool persistent;
typedef union { max_align_t alignment; size_t magic; } Header;
static void *observed_malloc(size_t n) {
    size_t at=calls++;
    if(at==fail_at||(persistent&&at>fail_at))return NULL;
    if(n>SIZE_MAX-sizeof(Header))return NULL;
    Header *h=malloc(sizeof *h+n);if(!h)return NULL;
    h->magic=0xabcdefu;live++;return h+1;
}
static void observed_free(void *p) {
    if(!p)return;
    Header *h=(Header *)p-1;assert(h->magic==0xabcdefu&&live);live--;h->magic=0;free(h);
}
#define malloc observed_malloc
#include "../src/module_sdk_abi.inc"
#undef malloc
static const char valid[]=
"{\"typed_abi\":{\"version\":1,\"target\":\"test-host\",\"types\":["
"{\"semantic\":\"int\",\"c_type\":\"int64_t\",\"kind\":\"int\"},"
"{\"semantic\":\"string\",\"c_type\":\"BorrowedString\",\"kind\":\"string\"},"
"{\"semantic\":\"Handle\",\"c_type\":\"HandlePtr\",\"kind\":\"opaque\"},"
"{\"semantic\":\"Packet\",\"c_type\":\"Packet\",\"kind\":\"record\",\"fields\":[\"value\",\"owner\"]},"
"{\"semantic\":\"(int,string)\",\"c_type\":\"Pair\",\"kind\":\"tuple\",\"fields\":[\"first\",\"second\"]},"
"{\"semantic\":\"Choice\",\"c_type\":\"Choice\",\"kind\":\"union\",\"tag\":\"kind\",\"variants\":[{\"tag\":\"CHOICE_A\",\"fields\":[\"data.a\"]},{\"tag\":\"CHOICE_B\",\"fields\":[]}]},"
"{\"semantic\":\"Mode\",\"c_type\":\"Mode\",\"kind\":\"enum\",\"variants\":[\"MODE_A\",\"MODE_B\"]},"
"{\"semantic\":\"array<Packet>\",\"c_type\":\"DynArray\",\"kind\":\"array\",\"abi\":\"dyn_array_v2\",\"element\":3},"
"{\"semantic\":\"fn(int)->int\",\"c_type\":\"Callback\",\"kind\":\"function\"}"
"],\"functions\":[{\"name\":\"read\",\"symbol\":\"provider_read\",\"parameters\":[2,7],\"result\":4},{\"name\":\"release\",\"symbol\":\"provider_release\",\"parameters\":[2],\"result\":null}]}}";
static void refusal(const char *text) {
    cJSON *j=cJSON_Parse(text);assert(j);ModuleBuildMetadata m={0};size_t baseline=live;
    assert(!module_parse_typed_abi(j,&m)&&!m.typed_abi&&live==baseline);cJSON_Delete(j);
}
int main(void) {
    cJSON_Hooks hooks={observed_malloc,observed_free};cJSON_InitHooks(&hooks);
    cJSON *j=cJSON_Parse(valid);assert(j);ModuleBuildMetadata m={0};size_t baseline=live;calls=0;
    assert(module_parse_typed_abi(j,&m)&&m.typed_abi);size_t measured=calls;
    char *saved=m.typed_abi;assert(!module_parse_typed_abi(j,&m)&&m.typed_abi==saved);
    cJSON_Delete(j);assert(strstr(saved,"provider_read"));
    cJSON *round=cJSON_CreateObject();cJSON *schema=cJSON_Parse(saved);assert(round&&schema&&cJSON_AddItemToObject(round,"typed_abi",schema));
    ModuleBuildMetadata again={0};assert(module_parse_typed_abi(round,&again)&&!strcmp(saved,again.typed_abi));
    observed_free(again.typed_abi);observed_free(saved);cJSON_Delete(round);assert(live==0);
    j=cJSON_Parse(valid);assert(j);baseline=live;
    for(unsigned mode=0;mode<2;mode++)for(size_t i=0;i<measured;i++) {
        calls=0;fail_at=i;persistent=mode!=0;m=(ModuleBuildMetadata){0};
        assert(!module_parse_typed_abi(j,&m)&&!m.typed_abi&&live==baseline);
        fail_at=SIZE_MAX;persistent=false;m=(ModuleBuildMetadata){0};
        assert(module_parse_typed_abi(j,&m));observed_free(m.typed_abi);assert(live==baseline);
    }
    cJSON_Delete(j);assert(live==0);
    j=cJSON_Parse("{\"name\":\"old\",\"unknown_old_field\":true}");m=(ModuleBuildMetadata){0};calls=0;
    assert(module_parse_typed_abi(j,&m)&&!m.typed_abi&&!calls);cJSON_Delete(j);
    const char *bad[]={
        "{\"typed_abi\":null}","{\"typed_abi\":{},\"typed_abi\":{}}",
        "{\"typed_abi\":{\"version\":1,\"version\":1,\"target\":\"x\",\"types\":[],\"functions\":[]}}",
        "{\"typed_abi\":{\"version\":2,\"target\":\"x\",\"types\":[],\"functions\":[]}}",
        "{\"typed_abi\":{\"version\":1,\"target\":\"x\",\"types\":[],\"functions\":[],\"extra\":0}}",
        "{\"typed_abi\":{\"version\":1,\"target\":\"x\",\"types\":[{\"semantic\":\"int\",\"c_type\":\"int;bad\",\"kind\":\"int\"}],\"functions\":[]}}",
        "{\"typed_abi\":{\"version\":1,\"target\":\"x\",\"types\":[{\"semantic\":\"A\",\"c_type\":\"A\",\"kind\":\"array\",\"abi\":\"dyn_array_v2\",\"element\":1}],\"functions\":[]}}",
        "{\"typed_abi\":{\"version\":1,\"target\":\"x\",\"types\":[{\"semantic\":\"A\",\"c_type\":\"A\",\"kind\":\"record\",\"fields\":[\"a..b\"]}],\"functions\":[]}}",
        "{\"typed_abi\":{\"version\":1,\"target\":\"x\",\"types\":[{\"semantic\":\"A\",\"c_type\":\"A\",\"kind\":\"record\",\"fields\":[\"a\",\"a\"]}],\"functions\":[]}}",
        "{\"typed_abi\":{\"version\":1,\"target\":\"x\",\"types\":[],\"functions\":[{\"name\":\"f\",\"symbol\":\"f\",\"parameters\":[0],\"result\":null}]}}"
    };
    for(size_t i=0;i<sizeof bad/sizeof bad[0];i++)refusal(bad[i]);
    assert(live==0);cJSON_InitHooks(NULL);
    puts("I passed owned typed ABI parsing, roundtrip and allocation rollback without admission.");return 0;
}
